### PR TITLE
OPENEUROPA-3378: Theme project content type top part

### PR DESCRIPTION
**Pull request by julien-**
### Description

Add top group, devided into two columns, first column containing body and results, the second column containing Details, Budget, Resources sub-groups.

### Change log

- Added:
  - oe_theme_content_project companion module
  - suggestions for pattern_field_list horizontal variant

### CODE REVIEW

File: modules/oe_theme_content_project/oe_theme_content_project.module
Comment (ademarco) [inline modules/oe_theme_content_project/oe_theme_content_project.module:34]: We need to test this with Behat, using our "oe_content Behat APIs", we first create a project with only start date, then we assert this label, then we add the end date, we refresh the page, and we assert the other label. The refreshing of the page is the important part, as it will allow us to correctly test caching invalidation.

File: modules/oe_theme_content_project/oe_theme_content_project.module
Comment (ademarco) [inline modules/oe_theme_content_project/oe_theme_content_project.module:25]: Could we move this just above the foreach below? :smile:

File: modules/oe_theme_content_project/oe_theme_content_project.module
Comment (ademarco) [inline modules/oe_theme_content_project/oe_theme_content_project.module]: Can we use here a Twig template? We can store them directly in the them /templatest/fields directory. It's always preferable, when possible, to use templates.

File: modules/oe_theme_content_project/src/Plugin/ExtraField/Display/PercentageExtraField.php
Comment (ademarco) [inline modules/oe_theme_content_project/src/Plugin/ExtraField/Display/PercentageExtraField.php]: Mh... don't we have any other way of aligning it with other fields? I'd rather not use empty labels for this.

File: modules/oe_theme_content_project/src/Plugin/ExtraField/Display/PercentageExtraField.php
Comment (ademarco) [inline modules/oe_theme_content_project/src/Plugin/ExtraField/Display/PercentageExtraField.php]: I think we can skip this, as the field's entity is always a render cache dependency of its fields, right @upchuk ?

File: modules/oe_theme_content_project/src/Plugin/ExtraField/Display/ProjectExtraFieldBase.php
Comment (ademarco) [inline modules/oe_theme_content_project/src/Plugin/ExtraField/Display/ProjectExtraFieldBase.php]: Do we need this class at all? It seems that only percentage field uses it, and it even doesn't use the entityTypeManager service.

File: modules/oe_theme_content_project/templates/field--node--oe-project-results--oe-project--full.html.twig
Comment (ademarco) [inline modules/oe_theme_content_project/templates/field--node--oe-project-results--oe-project--full.html.twig]: Since we are very specific in the template name (we know we are applying this only to a specific field, on a specific bundle, of a specific view mode) we can assume field cardinality etc. Let's make this template as simple as possible by taking advantages of those assumptions. Aslo, as we don't ship the full view mode, can we use this template instead?

```
field--node--oe-project-results--oe-project.html.twig 
```

File: modules/oe_theme_content_project/templates/pattern-field-list--variant-horizontal-oe-project.html.twig
Comment (ademarco) [inline modules/oe_theme_content_project/templates/pattern-field-list--variant-horizontal-oe-project.html.twig]: Can we use the field group template suggestions instead of creating one for the pattern? They should be already available to us.

File: oe_theme.theme
Comment (ademarco) [inline oe_theme.theme]: See above, if we use the field group template suggestions we don't need to implement this hook here.

**Review by ademarco — CHANGES_REQUESTED** [review]
Let's have this as DD/MM/YYYY:

image

Also remember to increase triple store to 1.3.0:
```
      image: openeuropa/triple-store-dev:1.3.0
```

I'm not sure what we did on event but I think we should ship with inline entity form here by default:

image

File: modules/oe_theme_content_project/oe_theme_content_project.module
Comment (julien-) [reply modules/oe_theme_content_project/oe_theme_content_project.module:34]: `oe_project` form display changes have been done on [oe_content PR221](https://github.com/openeuropa/oe_content/pull/221/files). Please @ademarco can you merge it?

File: modules/oe_theme_content_project/oe_theme_content_project.module
Comment (julien-) [reply modules/oe_theme_content_project/oe_theme_content_project.module:25]: Has it been fixed ? I do not understand your idea

File: modules/oe_theme_content_project/oe_theme_content_project.module
Comment (julien-) [reply modules/oe_theme_content_project/oe_theme_content_project.module]: The usage of template is now fixed.

Can you please explain me why not locate this templates in the `modules/oe_theme_content_project/templates/fields` ?
Thanks

File: modules/oe_theme_content_project/src/Plugin/ExtraField/Display/PercentageExtraField.php
Comment (julien-) [reply modules/oe_theme_content_project/src/Plugin/ExtraField/Display/PercentageExtraField.php]: Fixed using CSS

File: modules/oe_theme_content_project/src/Plugin/ExtraField/Display/ProjectExtraFieldBase.php
Comment (julien-) [reply modules/oe_theme_content_project/src/Plugin/ExtraField/Display/ProjectExtraFieldBase.php]: Fixed : Class removed and class `PercentageExtraField `extends now `ExtraFieldDisplayFormattedBase`

File: modules/oe_theme_content_project/src/Plugin/ExtraField/Display/PercentageExtraField.php
Comment (julien-) [reply modules/oe_theme_content_project/src/Plugin/ExtraField/Display/PercentageExtraField.php]: Removed but please @upchuk confirm that we dont need it. Thank you

File: modules/oe_theme_content_project/templates/field--node--oe-project-results--oe-project--full.html.twig
Comment (julien-) [reply modules/oe_theme_content_project/templates/field--node--oe-project-results--oe-project--full.html.twig]: Fxied : The file is renamed.

File: oe_theme.theme
Comment (julien-) [reply oe_theme.theme]: Fixed (made a commit to keep previous version)

File: modules/oe_theme_content_project/templates/pattern-field-list--variant-horizontal-oe-project.html.twig
Comment (julien-) [reply modules/oe_theme_content_project/templates/pattern-field-list--variant-horizontal-oe-project.html.twig]: Fixed (made a commit to keep previous version)

File: modules/oe_theme_content_project/oe_theme_content_project.module
Comment (ademarco) [reply modules/oe_theme_content_project/oe_theme_content_project.module:25]: You are using the `$label_overrides` in the foreach below and between this line and that foreach there is an if that it not related to either of those. It would make the whole block more readable by having:

```php
    $label_overrides = [
      'oe_project_dates' => t('Project duration'),
      'oe_project_website' => t('Project website'),
    ];
    foreach ($label_overrides as $name => $label_override) {
      if (!$entity->get($name)->isEmpty() && isset($build[$name]['#title'])) {
        $build[$name]['#title'] = $label_override;
      }
    }
```

File: modules/oe_theme_content_project/oe_theme_content_project.module
Comment (ademarco) [reply modules/oe_theme_content_project/oe_theme_content_project.module:34]: The linked PR was rejected, build was red + config had issues, please check.

File: composer.json
Comment (ademarco) [inline composer.json]: Use the branch at https://github.com/openeuropa/oe_content/pull/221 until that branch is merged in EPIC-Project.

File: modules/oe_theme_content_project/oe_theme_content_project.module
Comment (ademarco) [reply modules/oe_theme_content_project/oe_theme_content_project.module:34]: Also, no Behat tests are provided yet, we should add them before moving this into review.

File: modules/oe_theme_content_project/templates/field--node--oe-project-results--oe-project.html.twig
Comment (ademarco) [inline modules/oe_theme_content_project/templates/field--node--oe-project-results--oe-project.html.twig]: As I said before, when we override a template we don't need to account for all possible scenarios, especially when the field overrides are so specific as in this case. This can simply become:

```twig 
  <h2 class="ecl-u-type-heading-2">{{ label }}</h2>
  <div class="ecl-editor">
    {% for item in items %}
      {{ item.content }}
    {% endfor %}
  </div>
```

File: modules/oe_theme_content_project/templates/field--node--oe-project-results--oe-project.html.twig
Comment (ademarco) [inline modules/oe_theme_content_project/templates/field--node--oe-project-results--oe-project.html.twig]: As per other field templates, we just do:

```twig
{#
/**
 * @file
 * Theme override project results field template.
 *
 * @see ./core/themes/stable/templates/field/field.html.twig
 */
#}
```

File: templates/field/field--oe-project-departments.html.twig
Comment (ademarco) [inline templates/field/field--oe-project-departments.html.twig]: Same here, we know this is a multiple fields, so we can loose all extra markup, also let's scope this template down to node/bundle, as we did for the project results: `field--node--...--oe-project.html.twig`

File: templates/field/field--oe-project-website.html.twig
Comment (ademarco) [inline templates/field/field--oe-project-website.html.twig]: Same here, as comment above, simplify Twig markup taking into account that we know the field's cardinality and scope down template naming to node/bundle.

**Review by ademarco — CHANGES_REQUESTED** [review]

File: modules/oe_theme_content_project/oe_theme_content_project.module
Comment (yenyasinn) [reply modules/oe_theme_content_project/oe_theme_content_project.module:25]: It is correct order - default values are defined first, then we override one of them.

File: modules/oe_theme_content_project/oe_theme_content_project.module
Comment (yenyasinn) [reply modules/oe_theme_content_project/oe_theme_content_project.module:34]: Test is provided in `tests/features/project.feature`

File: sass/components/_content_project.scss
Comment (yenyasinn) [inline sass/components/_content_project.scss]: I called it _content_project because it is related to oe_theme _ **content_project** module.

File: tests/features/project.feature
Comment (drishu) [inline tests/features/project.feature]: Project period field has label "Start date" if start date is equal with end date

File: tests/features/project.feature
Comment (drishu) [inline tests/features/project.feature]: bit to much space on the right before the pipe

File: templates/content/field-group-html-element--group-result.html.twig
Comment (drishu) [inline templates/content/field-group-html-element--group-result.html.twig]: the if else here is not needed, Results will always be the title

File: modules/oe_theme_content_project/oe_theme_content_project.module
Comment (nagyad) [inline modules/oe_theme_content_project/oe_theme_content_project.module]: Please make this if into one line.

File: modules/oe_theme_content_project/src/Plugin/ExtraField/Display/PercentageExtraField.php
Comment (nagyad) [inline modules/oe_theme_content_project/src/Plugin/ExtraField/Display/PercentageExtraField.php]: EU acronym should be capital.

File: modules/oe_theme_content_project/src/Plugin/ExtraField/Display/PercentageExtraField.php
Comment (nagyad) [inline modules/oe_theme_content_project/src/Plugin/ExtraField/Display/PercentageExtraField.php]: This should be protected and not private.

File: sass/components/_content_project.scss
Comment (nagyad) [inline sass/components/_content_project.scss]: Please provide a comment header to explain what does this file contain in a sentence.
`/**`
` * @file`
` * This file is used to...`
 `*/`

File: tests/features/project.feature
Comment (nagyad) [inline tests/features/project.feature]: What does project tag do? Please remove if there is no associated logic to that.

File: tests/features/project.feature
Comment (nagyad) [inline tests/features/project.feature]: Sentences like this needs to end with a full stop.

File: tests/features/project.feature
Comment (nagyad) [inline tests/features/project.feature]: This empty line is not required.

**Review by nagyad — CHANGES_REQUESTED** [review]

File: modules/oe_theme_content_project/README.md
Comment (yenyasinn) [reply modules/oe_theme_content_project/README.md:14]: Looks like it was taken from `modules/oe_theme_content_event/README.md` - without this permission organisations in Participants and Coordinators fields won't be visible.

**Review by nagyad — DISMISSED** [review]
To be merged after https://github.com/openeuropa/oe_content/pull/221 is merged. Composer needs to be updated.

**Review by ademarco — DISMISSED** [review]

**Review by ademarco — DISMISSED** [review]

**Review by nagyad — DISMISSED** [review]

**Review by ademarco — APPROVED** [review]

File: tests/features/project.feature
Comment (ademarco) [inline tests/features/project.feature:2]: Since we are moving to browser tests to test the theme lets remove this and put it in "testProjectTopGroupRender()" test method above.

File: tests/Functional/ContentProjectRenderTest.php
Comment (ademarco) [inline tests/Functional/ContentProjectRenderTest.php:123]: Move behat test assertions here.

File: tests/Functional/ContentProjectRenderTest.php
Comment (ademarco) [inline tests/Functional/ContentProjectRenderTest.php:54]: Rename this in `testProjectRendering()`

**Review by ademarco — CHANGES_REQUESTED** [review]
\cc @yenyasinn

**Review by ademarco — APPROVED** [review]

**Comment by ademarco** [comment]
Approving changes as @yenyasinn will refactor the test in a follow up, which involves fixing tests too.

**Review by nagyad — APPROVED** [review]

**Pull request merged**
